File: frontend/src/api/algebra.ts
import { ApiError, apiUrl, fetchWithRetry, networkApiError, parseApiError } from './core';

export type AlgebraTopic = 'auto' | 'equation' | 'inequality' | 'exponential_log' | 'trigonometry' | 'complex' | 'sequence' | 'combinatorics_probability' | 'system' | 'parameter' | 'calculus_derivative' | 'calculus_limit' | 'calculus_integral';
export type AlgebraStatus = 'solved' | 'partial' | 'unsupported' | 'error';
export type AlgebraVerificationStatus = 'verified' | 'partially_verified' | 'failed' | 'skipped';
export type AlgebraInputFormat = 'auto' | 'plain' | 'latex' | 'structured';

export interface AlgebraSolveOptions {
  return_steps?: boolean;
  verify?: boolean;
  max_solutions?: number;
  prefer_exact?: boolean;
  grade_level?: 'C3';
  use_ai_extraction?: boolean;
  ai_explanation?: boolean;
}

export interface AlgebraInterval {
  variable?: string;
  start?: string | null;
  end?: string | null;
  closed_start?: boolean;
  closed_end?: boolean;
}

export type AlgebraAngleUnit = 'radian' | 'degree';

export interface AlgebraSolveRequest {
  input: string;
  input_format?: AlgebraInputFormat;
  topic?: AlgebraTopic;
  variables?: string[];
  parameters?: string[];
  domain?: 'R' | 'C' | 'N' | 'Z';
  angle_unit?: AlgebraAngleUnit;
  interval?: AlgebraInterval | null;
  options?: AlgebraSolveOptions;
}

export interface AlgebraInputChip {
  kind: 'intent' | 'format' | 'topic' | 'domain' | 'variable' | 'expression' | 'relation' | 'system';
  label: string;
  value: string;
}

export interface AlgebraInputInterpretation {
  detected_format: 'plain' | 'latex' | 'natural_vi' | 'mixed' | 'structured';
  source: 'raw' | 'rule_based_vi' | 'latex_normalizer' | 'structured_ui';
  canonical_input: string;
  topic_hint: AlgebraTopic;
  variables: string[];
  domain: 'R' | 'C' | 'N' | 'Z';
  chips: AlgebraInputChip[];
  warnings: string[];
}

export interface AlgebraSolutionValue {
  text: string;
  latex?: string | null;
  approximate?: string | null;
}

export interface AlgebraSolutionSet {
  kind: 'finite' | 'set' | 'interval' | 'periodic' | 'conditions' | 'expression' | 'empty' | 'unknown';
  text: string;
  latex?: string | null;
  values: AlgebraSolutionValue[];
}

export interface AlgebraVerificationCheck {
  name: string;
  status: 'pass' | 'fail' | 'warn' | 'skip';
  detail: string;
  latex?: string | null;
}

export interface AlgebraVerificationReport {
  status: AlgebraVerificationStatus;
  checks: AlgebraVerificationCheck[];
  method: string[];
}

export interface AlgebraSolveStep {
  index: number;
  title: string;
  explanation: string;
  short_explanation?: string | null;
  detail_level?: 'brief' | 'standard' | 'detailed';
  method?: string | null;
  goal?: string | null;
  why?: string | null;
  rule?: string | null;
  operation?: string | null;
  before_latex?: string | null;
  after_latex?: string | null;
  pitfall?: string | null;
  check?: string | null;
  expression?: string | null;
  expression_latex?: string | null;
  result?: string | null;
  result_latex?: string | null;
  kind?: 'normalize' | 'domain' | 'transform' | 'solve' | 'verify' | 'conclusion' | null;
  confidence?: 'verified' | 'symbolic' | 'numeric_checked' | 'unverified' | null;
  sub_steps?: AlgebraSolveStep[];
}

export interface AlgebraSolveResponse {
  input: string;
  normalized_input: string;
  input_interpretation?: AlgebraInputInterpretation | null;
  topic: string;
  problem_type: string;
  status: AlgebraStatus;
  answer: string;
  answer_latex?: string | null;
  solution_set: AlgebraSolutionSet;
  steps: AlgebraSolveStep[];
  verification: AlgebraVerificationReport;
  assumptions: string[];
  warnings: string[];
  errors: string[];
  request_id?: string | null;
  timings_ms?: Record<string, number>;
}

export async function solveAlgebra(payload: AlgebraSolveRequest): Promise<AlgebraSolveResponse> {
  try {
    const response = await fetchWithRetry(apiUrl('/api/algebra/solve'), {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      credentials: 'include',
      body: JSON.stringify(payload),
    });
    const body = await response.text();
    if (!body.trim()) throw new ApiError('Không thể giải bài đại số.');
    let parsed: unknown;
    try {
      parsed = JSON.parse(body);
    } catch {
      throw new ApiError('Không thể giải bài đại số.');
    }
    // 504 may still carry a structured AlgebraSolveResponse (request_id, warnings, timings).
    if (
      response.status === 504
      && parsed
      && typeof parsed === 'object'
      && 'status' in (parsed as object)
      && 'answer' in (parsed as object)
    ) {
      return parsed as AlgebraSolveResponse;
    }
    if (!response.ok) {
      throw await parseApiError(
        new Response(body, { status: response.status, headers: response.headers }),
        `Không thể giải bài đại số. HTTP ${response.status}`,
      );
    }
    return parsed as AlgebraSolveResponse;
  } catch (caught) {
    if (caught instanceof ApiError) throw caught;
    throw networkApiError(caught, 'Không thể giải bài đại số.');
  }
}

File: backend/app/schemas/algebra.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field, StrictBool

MAX_ALGEBRA_INPUT_CHARS = 2_000

AlgebraTopic = Literal[
    "auto",
    "equation",
    "inequality",
    "exponential_log",
    "trigonometry",
    "complex",
    "sequence",
    "combinatorics_probability",
    "statistics",
    "system",
    "parameter",
    "calculus_derivative",
    "calculus_derivative_by_definition",
    "calculus_limit",
    "calculus_continuous_at",
    "calculus_integral",
]
AlgebraStatus = Literal["solved", "partial", "unsupported", "error"]
VerificationStatus = Literal["verified", "partially_verified", "failed", "skipped"]
VerificationCheckStatus = Literal["pass", "fail", "warn", "skip"]
InputChipKind = Literal["intent", "format", "topic", "domain", "variable", "expression", "relation", "system"]


class AlgebraInterval(BaseModel):
    variable: str = "x"
    start: str | None = None
    end: str | None = None
    closed_start: bool = True
    closed_end: bool = True


class AlgebraSolveOptions(BaseModel):
    return_steps: bool = True
    verify: bool = True
    max_solutions: int = 50
    prefer_exact: bool = True
    grade_level: Literal["C3"] = "C3"
    use_ai_extraction: bool = False
    ai_explanation: bool = False


class AlgebraInputChip(BaseModel):
    kind: InputChipKind
    label: str
    value: str


class AlgebraInputInterpretation(BaseModel):
    detected_format: Literal["plain", "latex", "natural_vi", "mixed", "structured"] = "plain"
    source: Literal["raw", "rule_based_vi", "latex_normalizer", "structured_ui"] = "raw"
    canonical_input: str
    topic_hint: AlgebraTopic = "auto"
    variables: list[str] = Field(default_factory=list)
    domain: Literal["R", "C", "N", "Z"] = "R"
    chips: list[AlgebraInputChip] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)


class AlgebraSolveRequest(BaseModel):
    input: str = Field(min_length=1, max_length=MAX_ALGEBRA_INPUT_CHARS)
    input_format: Literal["auto", "plain", "latex", "structured"] = "auto"
    topic: AlgebraTopic = "auto"
    variables: list[str] = Field(default_factory=list, max_length=8)
    parameters: list[str] = Field(default_factory=list, max_length=8)
    domain: Literal["R", "C", "N", "Z"] = "R"
    # "default" = FE default R (AI may warn); "user" = explicit user choice (sticky).
    domain_source: Literal["default", "user"] = "default"
    # Radian is the only internal unit; degree rewrites trig args as x*pi/180.
    angle_unit: Literal["radian", "degree"] = "radian"
    interval: AlgebraInterval | None = None
    options: AlgebraSolveOptions = Field(default_factory=AlgebraSolveOptions)
    # When true and user is logged in, server persists a history row after solve.
    save_history: StrictBool = True


class AlgebraSolveStep(BaseModel):
    index: int
    title: str
    explanation: str
    short_explanation: str | None = None
    detail_level: Literal["brief", "standard", "detailed"] = "standard"
    method: str | None = None
    goal: str | None = None
    why: str | None = None
    rule: str | None = None
    operation: str | None = None
    before_latex: str | None = None
    after_latex: str | None = None
    pitfall: str | None = None
    check: str | None = None
    expression: str | None = None
    expression_latex: str | None = None
    result: str | None = None
    result_latex: str | None = None
    kind: Literal["normalize", "domain", "transform", "solve", "verify", "conclusion"] | None = None
    confidence: Literal["verified", "symbolic", "numeric_checked", "unverified"] | None = None
    sub_steps: list['AlgebraSolveStep'] = Field(default_factory=list)


class AlgebraSolutionValue(BaseModel):
    text: str
    latex: str | None = None
    approximate: str | None = None


class AlgebraSolutionSet(BaseModel):
    kind: Literal["finite", "set", "interval", "periodic", "conditions", "expression", "empty", "unknown"] = "unknown"
    text: str = ""
    latex: str | None = None
    values: list[AlgebraSolutionValue] = Field(default_factory=list)


class AlgebraVerificationCheck(BaseModel):
    name: str
    status: VerificationCheckStatus
    detail: str
    latex: str | None = None


class AlgebraVerificationReport(BaseModel):
    status: VerificationStatus = "skipped"
    checks: list[AlgebraVerificationCheck] = Field(default_factory=list)
    method: list[str] = Field(default_factory=list)


class AlgebraSolveResponse(BaseModel):
    input: str
    normalized_input: str
    input_interpretation: AlgebraInputInterpretation | None = None
    topic: str
    problem_type: str
    status: AlgebraStatus
    answer: str
    answer_latex: str | None = None
    solution_set: AlgebraSolutionSet = Field(default_factory=AlgebraSolutionSet)
    steps: list[AlgebraSolveStep] = Field(default_factory=list)
    milestones: list[str] = Field(default_factory=list)
    verification: AlgebraVerificationReport = Field(default_factory=AlgebraVerificationReport)
    assumptions: list[str] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)
    errors: list[str] = Field(default_factory=list)
    request_id: str | None = None
    timings_ms: dict[str, int] = Field(default_factory=dict)
    history_id: str | None = None
    cost_score: int | None = None


class AlgebraHistoryItem(BaseModel):
    id: str
    title: str | None = None
    problem_preview: str = ""
    topic: str = "auto"
    status: str = "solved"
    request_id: str | None = None
    is_favorite: bool = False
    archived_at: str | None = None
    created_at: str | None = None
    updated_at: str | None = None


class AlgebraHistoryDetail(AlgebraHistoryItem):
    request_json: dict = Field(default_factory=dict)
    response: AlgebraSolveResponse | None = None


class AlgebraHistoryCreateRequest(BaseModel):
    request: AlgebraSolveRequest
    response: AlgebraSolveResponse
    title: str | None = None


class AlgebraHistoryPatchRequest(BaseModel):
    title: str | None = None
    is_favorite: bool | None = None
    archive: bool | None = None


class AlgebraExportPdfRequest(BaseModel):
    response: AlgebraSolveResponse | None = None
    history_id: str | None = None
